File: ruby/lib/rayz/texture_map.rb
module Rayz
  # Simple PPM image loader for texture mapping
  class PPMImage
    attr_reader :width, :height, :pixels

    def initialize(width, height)
      @width = width
      @height = height
      @pixels = Array.new(height) { Array.new(width) }
    end

    # Get color at pixel (x, y)
    def pixel_at(x, y)
      return Color.new(red: 0, green: 0, blue: 0) if x < 0 || x >= @width || y < 0 || y >= @height
      @pixels[y][x]
    end

    # Set color at pixel (x, y)
    def set_pixel(x, y, color)
      @pixels[y][x] = color if x >= 0 && x < @width && y >= 0 && y < @height
    end

    # Load from PPM file (simple P3 format)
    def self.load_ppm(filename)
      lines = File.readlines(filename).map(&:strip).reject { |l| l.empty? || l.start_with?("#") }
      format = lines.shift

      raise "Unsupported PPM format: #{format}" unless format == "P3"

      width, height = lines.shift.split.map(&:to_i)
      max_color = lines.shift.to_i

      image = new(width, height)
      values = lines.join(" ").split.map(&:to_i)

      values.each_slice(3).with_index do |rgb, index|
        x = index % width
        y = index / width
        r = rgb[0] / max_color.to_f
        g = rgb[1] / max_color.to_f
        b = rgb[2] / max_color.to_f
        image.set_pixel(x, y, Color.new(red: r, green: g, blue: b))
      end

      image
    end
  end

  # Texture mapping pattern
  class TextureMap < Pattern
    attr_reader :image, :uv_map

    def initialize(image, uv_map)
      super()
      @image = image
      @uv_map = uv_map  # A proc that converts (point) -> [u, v] in range [0, 1]
    end

    def pattern_at(point)
      # Get UV coordinates from the mapping function
      u, v = @uv_map.call(point)

      # Convert UV to pixel coordinates
      # V is inverted because image coordinates start at top-left
      x = (u * (@image.width - 1)).round
      y = ((1.0 - v) * (@image.height - 1)).round

      @image.pixel_at(x, y)
    end

    # Planar mapping: maps (x, y) directly to (u, v)
    def self.planar_map
      ->(point) {
        u = point.x % 1.0
        v = point.z % 1.0
        [u, v]
      }
    end

    # Cylindrical mapping: wraps around Y axis
    def self.cylindrical_map
      ->(point) {
        # Compute angle around Y axis
        theta = Math.atan2(point.x, point.z)
        # Normalize to [0, 1]
        u = (theta + Math::PI) / (2 * Math::PI)
        # Y maps directly to v (assume unit cylinder 0 to 1)
        v = point.y % 1.0
        [u, v]
      }
    end

    # Spherical mapping: maps sphere surface to UV
    def self.spherical_map
      ->(point) {
        # Convert to spherical coordinates
        theta = Math.atan2(point.x, point.z)
        # Vector from center to point
        vec = Vector.new(x: point.x, y: point.y, z: point.z)
        radius = vec.magnitude
        phi = Math.acos(point.y / radius)

        # Map to UV coordinates
        u = 1.0 - (theta + Math::PI) / (2 * Math::PI)
        v = 1.0 - phi / Math::PI
        [u, v]
      }
    end
  end
end

File: python/rayz/texture_map.py
# python/rayz/texture_map.py
from __future__ import annotations

import math
from typing import Callable

from rayz.color import Color
from rayz.pattern import Pattern
from rayz.tuple import Point, Vector


class PPMImage:
    def __init__(self, width: int, height: int) -> None:
        self.width = width
        self.height = height
        self._pixels: list[list[Color | None]] = [[None] * width for _ in range(height)]

    def pixel_at(self, x: int, y: int) -> Color:
        if x < 0 or x >= self.width or y < 0 or y >= self.height:
            return Color(0, 0, 0)
        return self._pixels[y][x] or Color(0, 0, 0)

    def set_pixel(self, x: int, y: int, color: Color) -> None:
        if 0 <= x < self.width and 0 <= y < self.height:
            self._pixels[y][x] = color

    @classmethod
    def load_ppm(cls, filename: str) -> PPMImage:
        with open(filename) as f:
            raw = f.read()
        lines = [ln.strip() for ln in raw.splitlines() if ln.strip() and not ln.strip().startswith("#")]
        assert lines[0] == "P3", f"Unsupported PPM format: {lines[0]}"
        width, height = map(int, lines[1].split())
        max_color = int(lines[2])
        values = list(map(int, " ".join(lines[3:]).split()))
        image = cls(width, height)
        for i, (r, g, b) in enumerate(zip(values[::3], values[1::3], values[2::3])):
            image.set_pixel(i % width, i // width, Color(r / max_color, g / max_color, b / max_color))
        return image

    @classmethod
    def checkerboard(cls, width: int, height: int, checks: int) -> PPMImage:
        """Programmatic checkerboard — useful for demos without an image file."""
        img = cls(width, height)
        cell_w = max(1, width // checks)
        cell_h = max(1, height // checks)
        white = Color(1, 1, 1)
        black = Color(0.1, 0.1, 0.1)
        for y in range(height):
            for x in range(width):
                color = white if ((x // cell_w) + (y // cell_h)) % 2 == 0 else black
                img.set_pixel(x, y, color)
        return img


class TextureMap(Pattern):
    def __init__(self, image: PPMImage, uv_map: Callable) -> None:
        super().__init__()
        self.image = image
        self.uv_map = uv_map

    def pattern_at(self, point) -> Color:
        u, v = self.uv_map(point)
        x = round(u * (self.image.width - 1))
        y = round((1.0 - v) * (self.image.height - 1))
        return self.image.pixel_at(x, y)

    @staticmethod
    def planar_map(point) -> tuple[float, float]:
        return point.x % 1.0, point.z % 1.0

    @staticmethod
    def cylindrical_map(point) -> tuple[float, float]:
        theta = math.atan2(point.x, point.z)
        u = (theta + math.pi) / (2 * math.pi)
        return u, point.y % 1.0

    @staticmethod
    def spherical_map(point) -> tuple[float, float]:
        theta = math.atan2(point.x, point.z)
        radius = math.sqrt(point.x**2 + point.y**2 + point.z**2)
        phi = math.acos(point.y / radius) if radius > 0 else 0.0
        u = 1.0 - (theta + math.pi) / (2 * math.pi)
        v = 1.0 - phi / math.pi
        return u, v
